### PR TITLE
Handle empty request body

### PR DESCRIPTION
**Pull request by workmajj**
/wave @alexandrinaw @lgonzalez-silen @pascallouisperez

### CODE REVIEW

File: handlers.go [hunk 1]
```diff
@@ -46,12 +47,17 @@ type handlerMaker struct {
 	requestType, responseType reflect.Type
 }
 
+var emptyStructType = reflect.TypeOf(struct{}{})
```
Comment (pascallouisperez) [inline handlers.go]: What if you write `type foo struct{}`, wouldn't that be a different type than this? Wondering whether you'd be better off using `reflext` to do the type matching.

File: handlers_test.go [hunk 2]
```diff
@@ -42,14 +44,24 @@ func (_ *MurphySuite) TestHandlerOf(c *C) {
...
-	r.Body = ioutil.NopCloser(strings.NewReader(""))
+	r.Body = ioutil.NopCloser(strings.NewReader(`{"foo": not-a-string}`))
 
 	JsonHandler(correct)(w, r)
 
 	c.Assert(w.RecordedCode, Equals, http.StatusBadRequest)
 	c.Assert(w.RecordedBody, Equals, "{\"err\":\"unable to parse request\"}\n")
 }
 
+func (_ *MurphySuite) TestJsonHandler_unexpectedEmptyRequest(c *C) {
```
Comment (pascallouisperez) [inline handlers_test.go]: Can you also add a test for the case where you do expect an empty request?

**Comment by workmajj** [comment]
This should work better now, for any request struct that is empty regardless of defined type.

h/t @alexandrinaw for chatting with me about it yesterday 🙌

File: handlers_test.go [hunk 3]
```diff
@@ -19,14 +19,20 @@ type MurphySuite struct{}
...
+	Foo string `json:"foo"`
+}
 
 type sampleResponse struct{}
 
 func correct(ctx HttpContext, request *sampleRequest, response *sampleResponse) error {
 	return nil
 }
 
+func correctEmptyRequest(ctx HttpContext, request *struct{}, response *sampleResponse) error {
```
Comment (pascallouisperez) [inline handlers_test.go]: nit: instead of `request` use `_`

when I read this in the other PR, I thought this syntax was super neat to say "me no care 'bout your input"

Comment (workmajj) [reply handlers_test.go]: Whoops, forgot to change this when I copied the func. Thanks for catching!
